File: solutions/haskell/nucleotide-count/1/src/DNA.hs
module DNA (nucleotideCounts, Nucleotide(..)) where

import Data.Map (Map)
import qualified Data.Map as Map

data Nucleotide = A | C | G | T deriving (Eq, Ord, Show)

nucleotideCounts :: String -> Either String (Map Nucleotide Int)
nucleotideCounts xs = fmap countNucleotides $ traverse mapToNucleotide xs

countNucleotides :: [Nucleotide] -> Map Nucleotide Int
countNucleotides = foldl (\acc x -> Map.insert x (1 + (lookupOrZero x acc)) acc) Map.empty 

lookupOrZero :: Nucleotide -> Map Nucleotide Int -> Int
lookupOrZero x m = case Map.lookup x m of
                   Just n -> n
                   Nothing -> 0

mapToNucleotide :: Char -> Either String Nucleotide
mapToNucleotide 'A' = Right A
mapToNucleotide 'C' = Right C
mapToNucleotide 'G' = Right G
mapToNucleotide 'T' = Right T
mapToNucleotide _ = Left "Invalid strand"
                            

File: solutions/kotlin/nucleotide-count/1/src/main/kotlin/Dna.kt
class Dna(
    private val strand: String
) {

    init {
        if (strand.count { c -> c != 'A' && c != 'C' && c != 'G' && c != 'T'} > 0) {
            throw IllegalArgumentException("Invalid strand")
        }
    }
    
    val nucleotideCounts: Map<Char, Int>
        get() {
            val count = mutableMapOf<Char, Int>('A' to 0, 'C' to 0, 'G' to 0, 'T' to 0)
            strand.forEach { c ->
                count.put(c, count.getOrDefault(c, 0) + 1)
            }
            return count
        }
}
